org.eclipse.scanning.sequencer/src/org/eclipse/scanning/sequencer/nexus/NexusScanFileManager.java

```diff
@@ -244,62 +246,79 @@ public NexusScanInfo getNexusScanInfo() {
+				.map(m -> m.getName()).collect(Collectors.toSet());
+		perScanMonitorNames.addAll(existingPerScanMonitorNames);
 		
 		// add the global metadata scannables, and the required metadata scannables for
-		// each scannable in the scan
-		metadataScannableNames.addAll(deviceConnectorService.getGlobalMetadataScannableNames());
+		// each scannable in the scan. These are from the legacy GDA8 location map
+		perScanMonitorNames.addAll(scannableDeviceService.getGlobalMetadataScannableNames());
 		
 		// the set of scannable names to check for dependencies
 		Set<String> scannableNamesToCheck = new HashSet<>();
-		scannableNamesToCheck.addAll(metadataScannableNames);
+		scannableNamesToCheck.addAll(perScanMonitorNames);
 		scannableNamesToCheck.addAll(scannableNames);
 		do {
 			// check the given set of scannable names for dependencies
 			// each iteration checks the scannable names added in the previous one
-			Set<String> requiredScannables = scannableNamesToCheck.stream().flatMap(
-					name -> deviceConnectorService.getRequiredMetadataScannableNames(name).stream())
-					.filter(name -> !metadataScannableNames.contains(name))
+			Set<String> requiredScannables = scannableNamesToCheck.stream()
+					.flatMap(name -> scannableDeviceService.getRequiredMetadataScannableNames(name).stream())
+					.filter(name -> !perScanMonitorNames.contains(name))
 					.collect(Collectors.toSet());
 			
-			metadataScannableNames.addAll(requiredScannables);
+			perScanMonitorNames.addAll(requiredScannables);
 			scannableNamesToCheck = requiredScannables;
 		} while (!scannableNamesToCheck.isEmpty());
 		
-		// remove any scannable names in the scan from the list of metadata scannables
-		metadataScannableNames.removeAll(scannableNames);
-		
-		// get the metadata scannables for the given names
-		final List<IScannable<?>> monitors = model.getMonitors() != null ? model.getMonitors() : new ArrayList<>();
-		for (String scannableName : metadataScannableNames) {
-			IScannable<?> metadataScannable = deviceConnectorService.getScannable(scannableName);
-			try {
-			    metadataScannable.setMonitorRole(MonitorRole.PER_SCAN);
-			} catch (IllegalArgumentException ne) {
-				continue;
-			}
-			monitors.add(metadataScannable);
+		// remove any scannable names in the scan from the list of per scan monitors,
+		// as a scannable can only have one role within the scan
+		perScanMonitorNames.removeAll(scannableNames);
+		Set<String> scannablesToAdd = new HashSet<>(perScanMonitorNames.stream()
+				.filter(name -> !existingPerScanMonitorNames.contains(name))
+				.collect(Collectors.toSet()));
+		
+		// if there are any names of scannables to add, get the scannables for them,
+		// setting them to be per scan monitors
+		if (!scannablesToAdd.isEmpty()) {
+			final List<IScannable<?>> monitors = new ArrayList<>(model.getMonitors());
+			monitors.addAll(perScanMonitorNames.stream()
+				.map(name -> getPerScanMonitor(name)).collect(Collectors.toList()));
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Replace this lambda with a method reference. (sonar.java.source not set. Assuming 8 or greater.) [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS1612)